File: crates/mathhook-benchmarks/public/node/bench_mathhook.js
#!/usr/bin/env node
/**
 * MathHook Node.js Benchmark Suite
 *
 * Comprehensive polynomial operation benchmarks using MathHook Node.js bindings.
 * Public benchmarks with no external dependencies.
 *
 * Usage:
 *     node bench_mathhook.js [--json] [--iterations N]
 *
 * Output: JSON for baseline comparison or human-readable report.
 *
 * Last Updated: 2025-11-29T2000
 */

const path = require('path');
const fs = require('fs');

// Load MathHook Node.js bindings (search multiple paths)
let mathhook;
const possiblePaths = [
    path.join(__dirname, '../../../mathhook-node/index.js'),
    path.join(__dirname, '../../mathhook-node/index.js'),
    'mathhook-node'
];

for (const modulePath of possiblePaths) {
    try {
        mathhook = require(modulePath);
        break;
    } catch (e) {
        continue;
    }
}

if (!mathhook) {
    console.error('Error: MathHook Node.js bindings not found');
    console.error('Build with: cd crates/mathhook-node && npm install && npm run build');
    process.exit(1);
}

const { JsExpression, parse } = mathhook;

/**
 * Run benchmark with statistical analysis
 */
function benchmark(func, iterations = 100, warmup = 10) {
    // Warmup
    for (let i = 0; i < warmup; i++) {
        func();
    }

    // Collect timing samples (in nanoseconds)
    const times = [];
    for (let i = 0; i < iterations; i++) {
        const start = process.hrtime.bigint();
        func();
        const end = process.hrtime.bigint();
        times.push(Number(end - start));
    }

    // Statistics
    const sum = times.reduce((a, b) => a + b, 0);
    const mean_ns = sum / times.length;

    const variance = times.reduce((acc, t) => acc + Math.pow(t - mean_ns, 2), 0) / times.length;
    const stdev_ns = Math.sqrt(variance);

    times.sort((a, b) => a - b);
    const median_ns = times[Math.floor(times.length / 2)];
    const min_ns = times[0];
    const max_ns = times[times.length - 1];

    return {
        mean_ns,
        stdev_ns,
        median_ns,
        min_ns,
        max_ns,
        iterations
    };
}

// Parsing Benchmarks

function benchParseSimple() {
    return parse("x^2 + 2*x + 1");
}

function benchParseMedium() {
    return parse("x^5 + 3*x^4 - 2*x^3 + 7*x^2 - x + 5");
}

function benchParseLarge() {
    const terms = Array.from({length: 21}, (_, i) => `${i+1}*x^${i}`).join(" + ");
    return parse(terms);
}

function benchParseMultivariate() {
    return parse("x^2*y + x*y^2 + x*y + x + y + 1");
}

function benchParseLatex() {
    return parse("\\frac{x^2 + 1}{x - 1}");
}

// Polynomial Multiplication Benchmarks

function benchMulSimple() {
    const p1 = parse("x + 1");
    const p2 = parse("x + 1");
    return p1.multiply(p2);
}

function benchMulMedium() {
    const p1 = parse("x^10 + x^9 + x^8 + x^7 + x^6 + x^5 + x^4 + x^3 + x^2 + x + 1");
    const p2 = parse("2*x^10 + 2*x^9 + 2*x^8 + 2*x^7 + 2*x^6 + 2*x^5 + 2*x^4 + 2*x^3 + 2*x^2 + 2*x + 2");
    return p1.multiply(p2);
}

function benchMulLarge() {
    const terms = Array.from({length: 51}, (_, i) => `x^${i}`).join(" + ");
    const p1 = parse(terms);
    const p2 = parse(terms);
    return p1.multiply(p2);
}

function benchMulSparse() {
    const p1 = parse("1 + x^50 + x^100");
    const p2 = parse("1 + x^25 + x^75");
    return p1.multiply(p2);
}

// Polynomial Division Benchmarks

function benchDivSimple() {
    const p1 = parse("x^2 - 1");
    const p2 = parse("x - 1");
    return p1.divide(p2);
}

function benchDivMedium() {
    const p1 = parse("x^10 + x^5 + x^2 + 1");
    const p2 = parse("x^3 + x + 1");
    return p1.divide(p2);
}

// Expansion Benchmarks

function benchExpandSimple() {
    const p = parse("(x + 1)^2");
    return p.expand();
}

function benchExpandMedium() {
    const p = parse("(x + 1)^5");
    return p.expand();
}

function benchExpandLarge() {
    const p = parse("(x + 1)^10");
    return p.expand();
}

// Simplification Benchmarks

function benchSimplifySimple() {
    const p = parse("x + x + x");
    return p.simplify();
}

function benchSimplifyPolynomial() {
    const p = parse("x^2 + 2*x + 1 + x^2 - 1");
    return p.simplify();
}

function benchSimplifyLarge() {
    const p = parse("(x+1)^2 - (x^2 + 2*x + 1)");
    return p.simplify();
}

// Factorization Benchmarks

function benchFactorSimple() {
    const p = parse("x^2 - 1");
    return p.factor();
}

function benchFactorQuadratic() {
    const p = parse("x^2 + 2*x + 1");
    return p.factor();
}

// Main Benchmark Runner

function runAllBenchmarks(iterations = 100) {
    const benchmarks = {
        // Parsing
        parse_simple: benchParseSimple,
        parse_medium: benchParseMedium,
        parse_large: benchParseLarge,
        parse_multivariate: benchParseMultivariate,
        parse_latex: benchParseLatex,

        // Multiplication
        mul_simple: benchMulSimple,
        mul_medium: benchMulMedium,
        mul_large: benchMulLarge,
        mul_sparse: benchMulSparse,

        // Division
        div_simple: benchDivSimple,
        div_medium: benchDivMedium,

        // Expansion
        expand_simple: benchExpandSimple,
        expand_medium: benchExpandMedium,
        expand_large: benchExpandLarge,

        // Simplification
        simplify_simple: benchSimplifySimple,
        simplify_polynomial: benchSimplifyPolynomial,
        simplify_large: benchSimplifyLarge,

        // Factorization
        factor_simple: benchFactorSimple,
        factor_quadratic: benchFactorQuadratic,
    };

    const results = {
        platform: 'node-mathhook',
        binding: 'NAPI',
        node_version: process.version,
        benchmarks: {}
    };

    for (const [name, func] of Object.entries(benchmarks)) {
        try {
            process.stderr.write(`Running ${name}...\n`);
            const result = benchmark(func, iterations);
            results.benchmarks[name] = result;
        } catch (e) {
            process.stderr.write(`Error in ${name}: ${e.message}\n`);
            results.benchmarks[name] = { error: e.message };
        }
    }

    return results;
}

function printHumanReadable(results) {
    console.log('='.repeat(80));
    console.log('MathHook Node.js Benchmark Results');
    console.log('='.repeat(80));
    console.log(`Binding: ${results.binding}`);
    console.log(`Node version: ${results.node_version}`);
    console.log();

    const categories = {
        'Parsing': ['parse_'],
        'Multiplication': ['mul_'],
        'Division': ['div_'],
        'Expansion': ['expand_'],
        'Simplification': ['simplify_'],
        'Factorization': ['factor_'],
    };

    for (const [category, prefixes] of Object.entries(categories)) {
        console.log(`\n${category}:`);
        console.log('-'.repeat(60));

        for (const [name, data] of Object.entries(results.benchmarks)) {
            if (prefixes.some(prefix => name.startsWith(prefix))) {
                if (data.error) {
                    console.log(`  ${name.padEnd(30)} ERROR: ${data.error}`);
                } else {
                    const mean_us = data.mean_ns / 1000;
                    const stdev_us = data.stdev_ns / 1000;
                    console.log(`  ${name.padEnd(30)} ${mean_us.toFixed(2).padStart(12)} us  (stdev: ${stdev_us.toFixed(2).padStart(8)} us)`);
                }
            }
        }
    }
}

// CLI Handling

function main() {
    const args = process.argv.slice(2);
    let jsonOutput = false;
    let iterations = 100;

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--json') {
            jsonOutput = true;
        } else if (args[i] === '--iterations' && i + 1 < args.length) {
            iterations = parseInt(args[i + 1], 10);
            i++;
        }
    }

    const results = runAllBenchmarks(iterations);

    if (jsonOutput) {
        console.log(JSON.stringify(results, null, 2));
    } else {
        printHumanReadable(results);
    }
}

if (require.main === module) {
    main();
}

module.exports = { runAllBenchmarks, benchmark };

File: crates/mathhook-benchmarks/public/python/bench_mathhook.py
#!/usr/bin/env python3
"""
MathHook Python Benchmark Suite

Comprehensive polynomial operation benchmarks using MathHook Python bindings.

Usage:
    python bench_mathhook.py [--json] [--iterations N]

Output: JSON for baseline comparison or human-readable report.

Last Updated: 2025-12-28T1200
"""

import json
import sys
import time
import statistics
import argparse
from typing import Dict, Any, Callable

try:
    from mathhook import parse, symbol, symbols, gcd
except ImportError:
    print("Error: MathHook Python bindings not installed", file=sys.stderr)
    print("Install with: pip install mathhook", file=sys.stderr)
    sys.exit(1)


def benchmark(func: Callable, iterations: int = 100, warmup: int = 10) -> Dict[str, float]:
    """Run benchmark with statistical analysis."""
    # Warmup
    for _ in range(warmup):
        func()

    # Collect timing samples
    times = []
    for _ in range(iterations):
        start = time.perf_counter_ns()
        func()
        end = time.perf_counter_ns()
        times.append(end - start)

    # Statistics in nanoseconds
    mean_ns = statistics.mean(times)
    stdev_ns = statistics.stdev(times) if len(times) > 1 else 0
    median_ns = statistics.median(times)
    min_ns = min(times)
    max_ns = max(times)

    return {
        'mean_ns': mean_ns,
        'stdev_ns': stdev_ns,
        'median_ns': median_ns,
        'min_ns': min_ns,
        'max_ns': max_ns,
        'iterations': iterations
    }


# GCD expressions (pre-parsed for fair benchmarking)
_gcd_simple_f = parse("x^2 - 1")
_gcd_simple_g = parse("x - 1")

_gcd_medium_f = parse("x^5 + x^4 + x^3 + x^2 + x + 1")
_gcd_medium_g = parse("x^3 - 1")

_gcd_large_f = parse("x^20 - 1")
_gcd_large_g = parse("x^10 - 1")

_gcd_content_f = parse("6*x^3 + 12*x^2 + 18*x")
_gcd_content_g = parse("9*x^2 + 18*x + 27")

_gcd_bivariate_f = parse("x*y + x")
_gcd_bivariate_g = parse("x*y")

_gcd_trivariate_f = parse("x*y*z")
_gcd_trivariate_g = parse("x*y")

# Multiplication expressions
_mul_simple_f = parse("x + 1")
_mul_simple_g = parse("x + 1")

_mul_medium_f = parse("x^10 + x^9 + x^8 + x^7 + x^6 + x^5 + x^4 + x^3 + x^2 + x + 1")
_mul_medium_g = parse("2*x^10 + 2*x^9 + 2*x^8 + 2*x^7 + 2*x^6 + 2*x^5 + 2*x^4 + 2*x^3 + 2*x^2 + 2*x + 2")

_mul_large_terms = " + ".join([f"x^{i}" for i in range(51)])
_mul_large_f = parse(_mul_large_terms)
_mul_large_g = parse(_mul_large_terms)

_mul_sparse_f = parse("1 + x^50 + x^100")
_mul_sparse_g = parse("1 + x^25 + x^75")

# Division expressions
_div_simple_f = parse("x^2 - 1")
_div_simple_g = parse("x - 1")

_div_medium_f = parse("x^10 + x^5 + x^2 + 1")
_div_medium_g = parse("x^3 + x + 1")

# Expansion expressions
_expand_simple = parse("(x + 1)^2")
_expand_medium = parse("(x + 1)^5")
_expand_large = parse("(x + 1)^10")

# Simplification expressions
_simplify_simple = parse("x + x + x")
_simplify_polynomial = parse("x^2 + 2*x + 1 + x^2 - 1")
_simplify_large = parse("(x+1)^2 - (x^2 + 2*x + 1)")

# Factorization expressions
_factor_simple = parse("x^2 - 1")
_factor_quadratic = parse("x^2 + 2*x + 1")


# ============================================================================
# PARSING BENCHMARKS (measure parsing time - this is expected to include overhead)
# ============================================================================

def bench_parse_simple():
    """Parse simple polynomial from string."""
    return parse("x^2 + 2*x + 1")

def bench_parse_medium():
    """Parse medium complexity polynomial."""
    return parse("x^5 + 3*x^4 - 2*x^3 + 7*x^2 - x + 5")

def bench_parse_large():
    """Parse large polynomial (degree 20)."""
    terms = " + ".join([f"{i+1}*x^{i}" for i in range(21)])
    return parse(terms)

def bench_parse_multivariate():
    """Parse multivariate polynomial."""
    return parse("x^2*y + x*y^2 + x*y + x + y + 1")


# ============================================================================
# GCD BENCHMARKS (using pre-parsed expressions for fair comparison)
# ============================================================================

def bench_gcd_simple():
    """GCD of simple polynomials: (x^2-1, x-1) -> x-1"""
    return gcd(_gcd_simple_f, _gcd_simple_g)

def bench_gcd_medium():
    """GCD of degree-5 polynomials."""
    return gcd(_gcd_medium_f, _gcd_medium_g)

def bench_gcd_large():
    """GCD of degree-20 polynomials."""
    return gcd(_gcd_large_f, _gcd_large_g)

def bench_gcd_with_content():
    """GCD with common content factor."""
    return gcd(_gcd_content_f, _gcd_content_g)

def bench_gcd_bivariate():
    """Bivariate GCD."""
    return gcd(_gcd_bivariate_f, _gcd_bivariate_g)

def bench_gcd_trivariate():
    """Trivariate GCD."""
    return gcd(_gcd_trivariate_f, _gcd_trivariate_g)


# ============================================================================
# POLYNOMIAL MULTIPLICATION BENCHMARKS (using pre-parsed expressions)
# ============================================================================

def bench_mul_simple():
    """Multiply (x+1)(x+1)."""
    return _mul_simple_f * _mul_simple_g

def bench_mul_medium():
    """Multiply degree-10 polynomials."""
    return _mul_medium_f * _mul_medium_g

def bench_mul_large():
    """Multiply degree-50 polynomials."""
    return _mul_large_f * _mul_large_g

def bench_mul_sparse():
    """Multiply sparse polynomials."""
    return _mul_sparse_f * _mul_sparse_g


# ============================================================================
# POLYNOMIAL DIVISION BENCHMARKS (using pre-parsed expressions)
# ============================================================================

def bench_div_simple():
    """Divide (x^2-1) by (x-1)."""
    return _div_simple_f / _div_simple_g

def bench_div_medium():
    """Divide degree-10 by degree-3."""
    return _div_medium_f / _div_medium_g


# ============================================================================
# EXPANSION BENCHMARKS (using pre-parsed expressions)
# ============================================================================

def bench_expand_simple():
    """Expand (x+1)^2."""
    return _expand_simple.expand()

def bench_expand_medium():
    """Expand (x+1)^5."""
    return _expand_medium.expand()

def bench_expand_large():
    """Expand (x+1)^10."""
    return _expand_large.expand()


# ============================================================================
# SIMPLIFICATION BENCHMARKS (using pre-parsed expressions)
# ============================================================================

def bench_simplify_simple():
    """Simplify x + x + x."""
    return _simplify_simple.simplify()

def bench_simplify_polynomial():
    """Simplify polynomial expression."""
    return _simplify_polynomial.simplify()

def bench_simplify_large():
    """Simplify (x+1)^2 - (x^2 + 2*x + 1)."""
    return _simplify_large.simplify()


# ============================================================================
# FACTORIZATION BENCHMARKS (using pre-parsed expressions)
# ============================================================================

def bench_factor_simple():
    """Factor x^2 - 1."""
    return _factor_simple.factor()

def bench_factor_quadratic():
    """Factor x^2 + 2x + 1."""
    return _factor_quadratic.factor()


# ============================================================================
# MAIN BENCHMARK RUNNER
# ============================================================================

def run_all_benchmarks(iterations: int = 100) -> Dict[str, Any]:
    """Run all benchmarks and return results."""

    benchmarks = {
        # Parsing (with string processing overhead - this is expected)
        'parse_simple': bench_parse_simple,
        'parse_medium': bench_parse_medium,
        'parse_large': bench_parse_large,
        'parse_multivariate': bench_parse_multivariate,

        # GCD (using pre-parsed expressions for fair comparison)
        'gcd_simple': bench_gcd_simple,
        'gcd_medium': bench_gcd_medium,
        'gcd_large': bench_gcd_large,
        'gcd_with_content': bench_gcd_with_content,
        'gcd_bivariate': bench_gcd_bivariate,
        'gcd_trivariate': bench_gcd_trivariate,

        # Multiplication (using pre-parsed expressions)
        'mul_simple': bench_mul_simple,
        'mul_medium': bench_mul_medium,
        'mul_large': bench_mul_large,
        'mul_sparse': bench_mul_sparse,

        # Division (using pre-parsed expressions)
        'div_simple': bench_div_simple,
        'div_medium': bench_div_medium,

        # Expansion (using pre-parsed expressions)
        'expand_simple': bench_expand_simple,
        'expand_medium': bench_expand_medium,
        'expand_large': bench_expand_large,

        # Simplification (using pre-parsed expressions)
        'simplify_simple': bench_simplify_simple,
        'simplify_polynomial': bench_simplify_polynomial,
        'simplify_large': bench_simplify_large,

        # Factorization (using pre-parsed expressions)
        'factor_simple': bench_factor_simple,
        'factor_quadratic': bench_factor_quadratic,
    }

    results = {
        'platform': 'python-mathhook',
        'binding': 'PyO3',
        'python_version': sys.version.split()[0],
        'benchmarks': {}
    }

    for name, func in benchmarks.items():
        try:
            print(f"Running {name}...", file=sys.stderr)
            result = benchmark(func, iterations=iterations)
            results['benchmarks'][name] = result
        except Exception as e:
            print(f"Error in {name}: {e}", file=sys.stderr)
            results['benchmarks'][name] = {'error': str(e)}

    return results


def print_human_readable(results: Dict[str, Any]):
    """Print results in human-readable format."""
    print("=" * 80)
    print("MathHook Python Benchmark Results")
    print("=" * 80)
    print(f"Binding: {results['binding']}")
    print(f"Python version: {results['python_version']}")

    categories = {
        'Parsing': ['parse_'],
        'GCD': ['gcd_'],
        'Multiplication': ['mul_'],
        'Division': ['div_'],
        'Expansion': ['expand_'],
        'Simplification': ['simplify_'],
        'Factorization': ['factor_'],
    }

    for category, prefixes in categories.items():
        print(f"\n{category}:")
        print("-" * 60)

        for name, data in results['benchmarks'].items():
            if any(name.startswith(p) for p in prefixes):
                if 'error' in data:
                    print(f"  {name:30s} ERROR: {data['error']}")
                else:
                    mean_us = data['mean_ns'] / 1000
                    stdev_us = data['stdev_ns'] / 1000
                    print(f"  {name:30s} {mean_us:12.2f} us  (stdev: {stdev_us:8.2f} us)")


def main():
    parser = argparse.ArgumentParser(description='MathHook Python Benchmarks')
    parser.add_argument('--json', action='store_true', help='Output JSON format')
    parser.add_argument('--iterations', type=int, default=100, help='Benchmark iterations')
    args = parser.parse_args()

    results = run_all_benchmarks(iterations=args.iterations)

    if args.json:
        print(json.dumps(results, indent=2))
    else:
        print_human_readable(results)


if __name__ == '__main__':
    main()
